boilerplate/openshift/golang-osd-operator/csv-generate/common-generate-operator-bundle.py

```diff
@@ -41,25 +41,159 @@
+            f"{bindings}")
+
+def log_resource(resource):
+    """Log a message that we're processing the given resource.
+
+    :param resource: A k8s resource definition dict, expected to have at least
+        a kind and metadata.name.
+    """
+    print(f"Processing {resource['kind']} {resource['metadata']['name']}")
+
+def filename_for(resource):
+    """Generate a unique base file name for the given resource.
+
+    :param resource: A k8s resource definition dict, expected to have at least
+        a kind and metadata.name.
+    :return: A base file name (no directory) that should be unique for the
+        given resource (assuming the resource is itself unique).
+    """
+    return '_'.join((resource['kind'], resource['metadata']['name'])) + '.yaml'
+
+def bundle(resource):
+    """Create a yaml file for the given resource in the bundle.
+
+    Uses the VERSION_DIR global.
+
+    :param resource: A k8s resource definition dict, expected to have at least
+        a kind and metadata.name.
+    """
+    path = os.path.join(VERSION_DIR, filename_for(resource))
+    print(f"Bundling {path}")
+    with open(path, 'w') as outfile:
+        yaml.dump(resource, outfile, default_flow_style=False)
+
+def discover_service_account(binding):
+    """Discover or default the service account in a [Cluster]RoleBinding.
+
+    If binding is None, default to the operator name.
+    Uses the OPERATOR_NAME global.
+
+    :param binding: A [Cluster]RoleBinding resource dict.
+    :return: A string ServiceAccount name.
+    :raise NoServiceAccountSubjectInBinding: if binding does not contain a
+        ServiceAccount Subject.
+    """
+    if not binding:
+        print(
+            f"  Using default ServiceAccount ({OPERATOR_NAME}).")
+        return OPERATOR_NAME
+    for subject in binding.get('subjects', []):
+        if subject['kind'] == 'ServiceAccount':
+            sa = subject['name']
+            print(f"  Discovered ServiceAccount {sa}.")
+            return sa
+    raise NoServiceAccountSubjectInBinding(binding)
+
+def trim_index(index, kind, item):
+    """Removes one or all items of a given kind from the index.
+
+    :param index: Dict, keyed by kind, of k8s resources.
+    :param kind: String kind name to trim.
+    :param item: The item to trim. May be:
+        - A dict k8s resource. That specific resource is removed. If not found,
+          this function is a no-op.
+        - None. This function is a no-op.
+        - The string 'ALL'. The entire list for the given kind is removed.
+    """
+    if kind not in index:
+        return
+    if not item:
+        return
+    if item == 'ALL':
+        index.pop(kind, None)
+        return
+    # Find the specific resource to remove
+    for i in range(len(index[kind])):
+        if index[kind][i]['metadata']['name'] == item['metadata']['name']:
+            found = i
+            break
+    else:
+        # not found
+        return
```

Comment:
We expect there to only every be a single item with `.metadata.name` of matching `kind` I _think_ this is ok but it would fail silently.

Reply:
I think we're concerned that it would erroneously remove the wrong item. I'll add match by ns on L173.